src/lib/blueprint.ts

```diff
@@ -357,13 +366,80 @@ const createRoute = async (
     throw new Error(`Could not resolve name for route at ${path}`)
   }
 
+  const endpoints = await createEndpoints(path, pathItem, context)
+
   return {
     path: routePath,
     name,
-    namespace: namespace != null ? { path: namespace } : null,
-    endpoints: await createEndpoints(path, pathItem, context),
+    namespace:
+      namespace != null
+        ? {
+            path: namespace,
+            isDeprecated: false,
+            isUndocumented: false,
+          }
+        : null,
+    endpoints,
     subroutes: [],
+    isUndocumented: false,
+    isDeprecated: false,
+  }
+}
+
+const updateRoutesStatus = (routes: Route[]): Route[] => {
+  return routes.map((route) => {
+    const isRouteDeprecated = route.endpoints.every(
+      (endpoint) => endpoint.isDeprecated,
+    )
+    const isRouteUndocumented = route.endpoints.every(
+      (endpoint) => endpoint.isUndocumented,
+    )
+
+    return {
+      ...route,
+      isDeprecated: isRouteDeprecated,
+      isUndocumented: isRouteUndocumented,
+    }
+  })
+}
+
+const updateNamespaceStatus = (routes: Route[]): Route[] => {
+  // Group routes by namespace
+  const namespaceGroups = routes.reduce<Record<string, Route[]>>(
+    (acc, route) => {
+      if (route.namespace?.path == null) return acc
+
+      const namespacePath = route.namespace.path
+
+      if (acc[namespacePath] == null) {
+        acc[namespacePath] = []
+      }
+
+      acc[namespacePath].push(route)
+
+      return acc
+    },
+    {},
+  )
+
+  // Update namespace status based on routes
+  for (const routesInNamespace of Object.values(namespaceGroups)) {
+    const isNamespaceDeprecated = routesInNamespace.every(
+      (route) => route.isDeprecated,
+    )
+    const isNamespaceUndocumented = routesInNamespace.every(
+      (route) => route.isUndocumented,
+    )
+
+    for (const route of routesInNamespace) {
+      if (route.namespace != null) {
+        route.namespace.isDeprecated = isNamespaceDeprecated
+        route.namespace.isUndocumented = isNamespaceUndocumented
+      }
+    }
```

Comment:
Is this function mutating the input in place? Is there a way to do it with a simple `map`?

If we really need to mutate the input, the function so make that clear and not return it back, e.g., `updateRoutesWithNamespaceStatus(routes: Route[]): void`. If we factor out `getRoutesByNamespace` though, I think we can still do this update like the others, e.g, with a simple `routes.map(addNamespaceStatusToRoutes)`. Note that JS map gets the entire array as an arg in the callback, so `addNamespaceStatusToRoutes` can just call `getRoutesByNamespace` as needed.

Reply:
aff41ae738b1be51580075cb3b0820484bdf1b3a